### PR TITLE
Considerations and pre-cautions when adding new media formats

### LINKED ISSUE #171: Considerations and pre-cautions when adding new media formats
It occurred to us that adding a new format would most likely need consideration in a bunch of places on the platform, and we don't have any good guidelines on this when we were reviewing https://github.com/w3ctag/design-reviews/issues/495

Here is a brief list (there is probably more)

 - Accept headers
 - Anything that allows egress of a given raster/media stream
 - WebCodecs
 - Other media related APIs

### PR DESCRIPTION
**Pull request by kenchris**
Fixes #171

Co-Authored-By: Sangwhan "fish" Moon sangwhan@iki.fi

***
<a href="https://pr-preview.s3.amazonaws.com/kenchris/design-principles/pull/263.html" title="Last updated on Jan 27, 2021, 1:04 AM UTC (79dbede)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/263/f5a116e...kenchris:79dbede.html" title="Last updated on Jan 27, 2021, 1:04 AM UTC (79dbede)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -1975,6 +1974,37 @@ offers guidance that should be considered in the development of new features, no
...
+There are cases when a new capability on the web involves adding a new data format.
+This can be an image, video, audio, or any other type of data that a browser is expected
+to ingest. New formats should have a standardized MIME type, which is strictly validated.
+
+While legacy media formats do not always have strict enforcement for MIME types (and
+sometimes rely on peeking at headers, to workaround this), this is mostly for legacy
+compatibility reasons and should not be expected or implemented for new formats.
+
+It is expected that spec authors also integrate the new format to existing APIs, so that
+they are whitelisted in both ingress (e.g. decoding from a ReadableStream) and egress
```
Comment (annevk) [inline index.bs:1991]:
```suggestion
they are safelisted in both ingress (e.g. decoding from a ReadableStream) and egress
```

File: index.bs [hunk 2]
```diff
@@ -1975,6 +1974,37 @@ offers guidance that should be considered in the development of new features, no
...
+encoder) for said image format to support decoding in HTMLImageElements. On top of this,
+you are also expected to add support to egress points such as HTMLCanvasElement.toBlob()
+and HTMLCanvasElement.toDataURL().
+
+For legacy reasons browsers support MIME type sniffing, but we do not recommend extending
+the [pattern matching algorithm](https://mimesniff.spec.whatwg.org/#image-type-pattern-matching-algorithm),
+and instead recommend enforcing strict MIME types for newer formats.
+
+New MIME types should be submitted to the Internet Engineering Steering Group (IESG) for review,
+approval, and registration with the Internet Assigned Numbers Authority (IANA).
```
Comment (annevk) [inline index.bs:2005]: I think they only need to be registered.

File: index.bs [hunk 3]
```diff
@@ -1975,6 +1974,37 @@ offers guidance that should be considered in the development of new features, no
...
+
+For example. if you are to add an image format to the web platform, first add a new MIME
+type for the format. After this, you would naturally add a decoder (and presumably an
+encoder) for said image format to support decoding in HTMLImageElements. On top of this,
+you are also expected to add support to egress points such as HTMLCanvasElement.toBlob()
+and HTMLCanvasElement.toDataURL().
+
+For legacy reasons browsers support MIME type sniffing, but we do not recommend extending
+the [pattern matching algorithm](https://mimesniff.spec.whatwg.org/#image-type-pattern-matching-algorithm),
+and instead recommend enforcing strict MIME types for newer formats.
```
Comment (annevk) [inline index.bs:2002]: Might mention the word security here.

File: index.bs (hunk 2 above)
Comment (kenchris) [reply index.bs:2005]: so just "should be registered with IANA"?

File: index.bs [hunk 4]
```diff
@@ -1975,6 +1974,37 @@ offers guidance that should be considered in the development of new features, no
...
```
Comment (kenchris) [inline index.bs:2002]:
```suggestion
For legacy reasons browsers support MIME type sniffing, but we do not recommend extending
the [pattern matching algorithm](https://mimesniff.spec.whatwg.org/#image-type-pattern-matching-algorithm),
due to security implications, and instead recommend enforcing strict MIME types for newer formats.
```

File: index.bs [hunk 5]
```diff
@@ -1975,6 +1974,37 @@ offers guidance that should be considered in the development of new features, no
...
+
+New MIME types should be submitted to the Internet Engineering Steering Group (IESG) for review,
+approval, and registration with the Internet Assigned Numbers Authority (IANA).
```
Comment (kenchris) [inline index.bs:2005]:
```suggestion
New MIME types should be registered with the Internet Assigned Numbers Authority (IANA).
```

File: index.bs (hunk 4 above)
Comment (kenchris) [reply index.bs:2002]: @cynthia ?

File: index.bs (hunk 5 above)
Comment (plinss) [reply index.bs:2005]: Maybe: "New MIME types should be registered with the Internet Assigned Numbers Authority (IANA) and the data format should have a publicly available specification."

File: index.bs [hunk 6]
```diff
@@ -1975,6 +1974,36 @@ offers guidance that should be considered in the development of new features, no
...
+type for the format. After this, you would naturally add a decoder (and presumably an
+encoder) for said image format to support decoding in HTMLImageElements. On top of this,
+you are also expected to add support to egress points such as HTMLCanvasElement.toBlob()
+and HTMLCanvasElement.toDataURL().
+
+For legacy reasons browsers support MIME type sniffing, but we do not recommend extending
+the [pattern matching algorithm](https://mimesniff.spec.whatwg.org/#image-type-pattern-matching-algorithm),
+due to security implications, and instead recommend enforcing strict MIME types for newer formats.
+
+New MIME types should be registered with the Internet Assigned Numbers Authority (IANA).
```
Comment (cynthia) [inline index.bs:2004]:
```suggestion
New MIME types should have a specification and should be registered with the Internet Assigned Numbers Authority (IANA).
```

**Review by cynthia — APPROVED** [review]

**Review by plinss — APPROVED** [review]

**Pull request merged**